libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/MapboxNavigationNotification.kt

```diff
@@ -0,0 +1,380 @@
+import com.mapbox.services.android.navigation.R
+import com.mapbox.services.android.navigation.v5.navigation.notification.NavigationNotification
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress
+import com.mapbox.services.android.navigation.v5.utils.DistanceFormatter
+import com.mapbox.services.android.navigation.v5.utils.LocaleUtils
+import com.mapbox.services.android.navigation.v5.utils.time.TimeFormatter.formatTime
+import java.util.Calendar
+
+/**
+ * This is in charge of creating the persistent navigation session notification and updating it.
+ */
+internal class MapboxNavigationNotification : NavigationNotification {
+
+    private val END_NAVIGATION_ACTION = "com.mapbox.intent.action.END_NAVIGATION"
+    private val SET_BACKGROUND_COLOR = "setBackgroundColor"
+
+    private var notificationManager: NotificationManager? = null
+    private var notification: Notification? = null
+    private var collapsedNotificationRemoteViews: RemoteViews? = null
+    private var expandedNotificationRemoteViews: RemoteViews? = null
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private var currentDistanceText: SpannableString? = null
+    private var distanceFormatter: DistanceFormatter? = null
+    private var instructionText: String? = null
+    private var currentManeuverId: Int = 0
+    private var isTwentyFourHourFormat: Boolean = false
+    private var etaFormat: String = ""
+    private val applicationContext: Context
+    private var pendingOpenIntent: PendingIntent? = null
+    private var pendingCloseIntent: PendingIntent? = null
+
+    private val endNavigationBtnReceiver = object : BroadcastReceiver() {
+        override fun onReceive(applicationContext: Context, intent: Intent) {
+            this@MapboxNavigationNotification.onEndNavigationBtnClick()
+        }
+    }
+
+    constructor(applicationContext: Context, mapboxNavigation: MapboxNavigation) {
+        this.applicationContext = applicationContext
+        initialize(applicationContext, mapboxNavigation)
+    }
+
+    // For testing only
+    constructor(
+        applicationContext: Context,
+        mapboxNavigation: MapboxNavigation,
+        notification: Notification
+    ) {
+        this.applicationContext = applicationContext
+        this.notification = notification
+        initialize(applicationContext, mapboxNavigation)
+    }
+
+    override fun getNotification() = notification
+
+    override fun getNotificationId(): Int {
+        return NavigationConstants.NAVIGATION_NOTIFICATION_ID
+    }
+
+    override fun updateNotification(routeProgress: RouteProgress) {
+        updateNotificationViews(routeProgress)
+        rebuildNotification()
+    }
+
+    override fun onNavigationStopped(applicationContext: Context) {
+        unregisterReceiver(applicationContext)
+    }
+
+    // Package private (no modifier) for testing purposes
+    fun generateArrivalTime(routeProgress: RouteProgress, time: Calendar): String? =
```

Comment:
Since this is kotlin doesn't it makes it public?

Reply:
It does. The tests will fail if this is made private

Reply:
@olegzil we should not break encapsulation for the sake of passing test, is there a way to preserve encapsulation while make it still testable?

Reply:
The converted code is different from the original in that Kotlin does not have the concept of Package-Private declarations. There are multiple annotations, i.e. @TestOnly and @VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE), but neither duplicates the effect of Java's Package-Private. This issue is going to come up in many other files. Kotlin does not provide a solution. The only thing that can be done is to rearchitect the class, but that approach will introduces potential regression not to mention require more time

Reply:
exposing internal methods as public as a result of conversion isn't a viable solution, it's important not to regress on encapsulation.

Reply:
This is exactly where [`libandroid-navigation-internal`](https://github.com/mapbox/mapbox-navigation-android/tree/libandroid-navigation-internal) work comes in handy 😬

Reply:
@Guardiola31337 it sounds like `libandroid-navigation-internal` might not be ideal long term. we should likely modularize our code to preserve encapsulation.

Reply:
It's a pretty common practice / "standard" in SDKs / libraries development, don't understand why might not work in the long term